bitcoin-tx: Require that input amount is provided for witness transactions

theStack opened the pull request:
This PR picks up the obviously abandoned PR #13608 (last activity was three and a half years ago) by rebasing it on master and adding missing tests. Original PR description: "_Applies fix from #12458 / #13547 to bitcoin-tx._"

The private key is the compressed version of the one used in most other util tests (5HpHagT65TZzG1PH3CSu63k8DbpvD8s5ip4nEB3kEsreAnchuDf, corresponds to the scalar value k=1 in big endian), since segwit signing refuses uncompressed keys.

The error message from the picked up PR is changed to not include the amount, as showing any value would be just confusing.

Empact committed:
Require that input amount is provided for bitcoin-tx witness transactions

Co-authored-by: Sebastian Falbesoner <sebastian.falbesoner@gmail.com>

theStack committed:
test: check that bitcoin-tx detects missing input amount for segwit transactions

josibake commented on src/bitcoin-tx.cpp:
```diff
@@ -669,6 +669,10 @@ static void MutateTxSign(CMutableTransaction& tx, const std::string& flagStr)
             ProduceSignature(keystore, MutableTransactionSignatureCreator(&mergedTx, i, amount, nHashType), prevPubKey, sigdata);
 
+        if (amount == MAX_MONEY && !sigdata.scriptWitness.IsNull()) {
+            throw std::runtime_error(strprintf("Missing amount for %s", coin.out.ToString()));
```
I tend to agree that seeing `CTxOut(nValue=21000000.00000000...` when the error is for a missing amount is confusing. 

Using `MAX_AMOUNT` for the check is already a nuanced concept (I had to read through all the linked PRs/issues before I understood why it was being used here). I don't think displaying 0 is correct, either, since a 0 amount UTXO is a valid spend.

Perhaps just print the `scriptPubKey`? Something like `Missing amount for CTxOut with scriptPubKey=0014751e76e8199196d454941c45d1`

theStack replied:
Thanks, I agree that the best would be to not display the amount at all if it is missing and picked up your suggestion.

josibake reviewed (commented):
Concept ACK

+1 for adding the tests! I don't think printing `nValue=21000000.00000000` is a great idea, so I left a suggestion inline

josibake commented:
One other suggestion: add a test with a 0 amount as that is a good (edge?) case to have tested.

theStack commented:
Thanks for the valuable review @josibake! Rebased on master and force-pushed with the following adaptions:
- changed the error message to not include the amount, as showing any value would be just confusing (https://github.com/bitcoin/bitcoin/pull/23784#discussion_r776366558)
- changed the input value of the first introduced passing test from 31337 to 0, as this is an edge-case (https://github.com/bitcoin/bitcoin/pull/23784#issuecomment-1002635962)

(`git range-diff 8dc5fa1...8bd34dc`)

josibake commented:
ACK https://github.com/bitcoin/bitcoin/commit/8bd34dc774788cbf3cad8e139542a0ed9f3e8bb4

MarcoFalke commented:
Needs OP adjusted before merge?

theStack commented:
> Needs OP adjusted before merge?

Thanks, done.
